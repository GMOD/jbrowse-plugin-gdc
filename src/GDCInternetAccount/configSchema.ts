import { ConfigurationSchema } from '@jbrowse/core/configuration'
import { Instance } from 'mobx-state-tree'
import PluginManager from '@jbrowse/core/PluginManager'
import { createInternetAccountConfig } from '@jbrowse/core/pluggableElementTypes/models'

function GDCConfigFactory(pluginManager: PluginManager) {
  return ConfigurationSchema(
    'GDCInternetAccount',
    {
      validDomains: {
        description:
          'array of valid domains the url can contain to use this account. Empty = all domains',
        type: 'stringArray',
        defaultValue: [],
      },
      authHeader: {
        description: 'custom auth header for authorization',
        type: 'string',
        defaultValue: 'Authorization',
      },
      customEndpoint: {
        description: 'custom endpoint for the external token resource',
        type: 'string',
        defaultValue: '',
      },
    },
    {
      baseConfiguration: createInternetAccountConfig(pluginManager),
      explicitlyTyped: true,
    },
  )
}

export type GDCInternetAccountConfigModel = ReturnType<typeof GDCConfigFactory>
export type OAuthInternetAccountConfig = Instance<GDCInternetAccountConfigModel>
export default GDCConfigFactory
