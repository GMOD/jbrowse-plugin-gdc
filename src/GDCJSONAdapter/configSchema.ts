import { ConfigurationSchema } from '@jbrowse/core/configuration'
import { types } from 'mobx-state-tree'

export default ConfigurationSchema(
  'GDCJSONAdapter',
  {
    featureType: {
      type: 'stringEnum',
      model: types.enumeration('Feature Type', ['mutation', 'gene']),
      defaultValue: 'mutation',
      description: 'The type of the track to add',
    },
    data: {
      type: 'string',
      defaultValue: '',
      description: 'JSON string of the GDC data from the file'
    }
  },
  { explicitlyTyped: true },
)
