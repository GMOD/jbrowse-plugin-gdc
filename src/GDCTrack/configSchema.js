import { ConfigurationSchema } from '@gmod/jbrowse-core/configuration'

export default pluginManager => {
    const { BaseTrackConfig: LinearGenomeTrackConfig }  = pluginManager.getPlugin('LinearGenomeViewPlugin').exports
    return ConfigurationSchema(
    'GDCTrack',
    {
      adapter: pluginManager.pluggableConfigSchemaType('adapter'),
      renderer: pluginManager.pluggableConfigSchemaType('renderer'),
    },
    { baseConfiguration: LinearGenomeTrackConfig, explicitlyTyped: true },
  )
}
