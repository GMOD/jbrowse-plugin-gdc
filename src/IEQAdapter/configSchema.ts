import { ConfigurationSchema } from '@jbrowse/core/configuration'

export default ConfigurationSchema(
  'IeqAdapter',
  {
    ieqLocation: {
      type: 'fileLocation',
      defaultValue: { uri: '/path/to/myfile.tsv', locationType: 'UriLocation' },
    },
  },
  { explicitlyTyped: true },
)
