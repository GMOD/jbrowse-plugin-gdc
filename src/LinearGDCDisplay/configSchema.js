import { ConfigurationSchema } from '@jbrowse/core/configuration'

export default pluginManager => {
  const { baseLinearDisplayConfigSchema } = pluginManager.getPlugin(
    'LinearGenomeViewPlugin',
  ).exports
  return ConfigurationSchema(
    'LinearGDCDisplay',
    { renderer: pluginManager.pluggableConfigSchemaType('renderer') },
    { baseConfiguration: baseLinearDisplayConfigSchema, explicitlyTyped: true },
  )
}
