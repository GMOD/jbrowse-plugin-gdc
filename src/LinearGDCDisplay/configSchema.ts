import PluginManager from '@jbrowse/core/PluginManager'
import { ConfigurationSchema } from '@jbrowse/core/configuration'
import LinearGenomeViewPlugin from '@jbrowse/plugin-linear-genome-view'

export default (pluginManager: PluginManager) => {
  const { baseLinearDisplayConfigSchema } = (
    pluginManager.getPlugin('LinearGenomeViewPlugin') as LinearGenomeViewPlugin
  ).exports
  return ConfigurationSchema(
    'LinearGDCDisplay',
    { renderer: pluginManager.pluggableConfigSchemaType('renderer') },
    { baseConfiguration: baseLinearDisplayConfigSchema, explicitlyTyped: true },
  )
}
