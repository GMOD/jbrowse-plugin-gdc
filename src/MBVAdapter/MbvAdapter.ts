import {
  BaseFeatureDataAdapter,
  BaseOptions,
} from '@jbrowse/core/data_adapters/BaseAdapter'
import { FileLocation, Region } from '@jbrowse/core/util/types'
import { openLocation } from '@jbrowse/core/util/io'
import { ObservableCreate } from '@jbrowse/core/util/rxjs'
import MbvFeature from './MbvFeature'
import { Feature } from '@jbrowse/core/util/simpleFeature'
import { readConfObject } from '@jbrowse/core/configuration'
import pako from 'pako'

export default class MbvAdapter extends BaseFeatureDataAdapter {
  public static capabilities = ['getFeatures', 'getRefNames']

  private setupP?: Promise<Feature[]>

  private readMbv(fileContents: string) {
    const lines = fileContents.split('\n')
    const refNames: string[] = []
    const rows: string[] = []
    let columns: string[] = []
    let refNameColumnIndex = 0
    lines.forEach(line => {
      if (columns.length === 0) {
        columns = line.split('\t')
        refNameColumnIndex = columns.findIndex(
          element => element.toLowerCase() === 'chromosome',
        )
      } else {
        if (
          line.split('\t')[refNameColumnIndex] !== '*' &&
          line.split('\t')[refNameColumnIndex] !== undefined
        ) {
          rows.push(line)
          refNames.push(line.split('\t')[refNameColumnIndex])
        }
      }
    })

    return {
      lines: rows,
      columns,
      refNames: Array.from(new Set(refNames)),
    }
  }

  private parseLine(line: string, columns: string[]) {
    const mutationObject: Record<string, unknown> = {}
    line.split('\t').forEach((property: string, i: number) => {
      if (property) {
        mutationObject[columns[i].toLowerCase()] = property
      }
    })
    return mutationObject
  }

  private async decodeFileContents() {
    const mbvLocation = readConfObject(
      this.config,
      'mbvLocation',
    ) as FileLocation

    const fileContents = await openLocation(
      mbvLocation,
      this.pluginManager,
    ).readFile()

    let str: string
    if (
      typeof fileContents[0] === 'number' &&
      fileContents[0] === 31 &&
      typeof fileContents[1] === 'number' &&
      fileContents[1] === 139 &&
      typeof fileContents[2] === 'number' &&
      fileContents[2] === 8
    ) {
      str = new TextDecoder().decode(pako.inflate(fileContents))
    } else {
      str = fileContents.toString()
    }

    return this.readMbv(str)
  }

  private async getLines() {
    const { columns, lines } = await this.decodeFileContents()

    return lines.map((line, index) => {
      return new MbvFeature({
        value: this.parseLine(line, columns),
        id: `${this.id}-mbv-${index}`,
      })
    })
  }

  private async setup() {
    if (!this.setupP) {
      this.setupP = this.getLines()
    }
    return this.setupP
  }

  public async getRefNames(_: BaseOptions = {}) {
    const { refNames } = await this.decodeFileContents()
    return refNames
  }

  public getFeatures(region: Region, opts: BaseOptions = {}) {
    return ObservableCreate<Feature>(async observer => {
      const feats = await this.setup()
      feats.forEach(f => {
        if (
          f.get('refName') === region.refName &&
          f.get('end') > region.start &&
          f.get('start') < region.end
        ) {
          observer.next(f)
        }
      })
      observer.complete()
    }, opts.signal)
  }

  public freeResources(): void {}
}
