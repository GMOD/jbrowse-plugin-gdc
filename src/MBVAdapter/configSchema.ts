import { ConfigurationSchema } from '@jbrowse/core/configuration'

export default ConfigurationSchema(
  'MbvAdapter',
  {
    mbvLocation: {
      type: 'fileLocation',
      defaultValue: { uri: '/path/to/myfile.txt', locationType: 'UriLocation' },
    },
  },
  { explicitlyTyped: true },
)
