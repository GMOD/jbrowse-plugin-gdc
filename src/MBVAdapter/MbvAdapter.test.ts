import { toArray } from 'rxjs/operators'
import MbvAdapter from './MbvAdapter'
import configSchema from './configSchema'
import fetchMock from 'jest-fetch-mock'
import { firstValueFrom } from 'rxjs'

fetchMock.enableMocks()

test('adapter can fetch features from mbv_test_data.txt', async () => {
  const adapter = new MbvAdapter(
    configSchema.create({
      mbvLocation: {
        localPath: require.resolve('./test_data/mbv_test_data.txt'),
      },
    }),
  )

  const features = adapter.getFeatures({
    assemblyName: 'hg38',
    refName: 'chr3',
    start: 0,
    end: 200000,
  })

  const names = await adapter.getRefNames()
  expect(names).toMatchSnapshot()

  const featuresArray = await firstValueFrom(features.pipe(toArray()))
  const featuresJsonArray = featuresArray.map(f => f.toJSON())
  expect(featuresJsonArray).toMatchSnapshot()
})
