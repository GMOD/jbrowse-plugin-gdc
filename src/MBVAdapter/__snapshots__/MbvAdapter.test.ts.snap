// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`adapter can fetch features from mbv_test_data.txt 1`] = `
[
  "chr16",
  "chr3",
  "chr7",
  "chr2",
  "chr11",
  "chr14",
  "chr18",
  "chr4",
  "chr1",
  "chr6",
  "chr22",
  "chr12",
  "chr17",
  "chr20",
  "chr10",
  "chrX",
  "chr19",
  "chr9",
  "chr13",
  "chr8",
  "chr5",
  "chr15",
  "chr21",
  "chrY",
]
`;

exports[`adapter can fetch features from mbv_test_data.txt 2`] = `
[
  {
    "beta_value": "0.854812485786774",
    "cgi_coordinate": "CGI:chr3:196708-198323",
    "chromosome": "chr3",
    "composite element ref": "cg00903242",
    "end": 196479,
    "feature_type": "N_Shore",
    "gene_symbol": "CHL1;CHL1;CHL1;CHL1;CHL1;CHL1;CHL1;CHL1-AS2;CHL1-AS2;CHL1-AS2",
    "gene_type": "protein_coding;protein_coding;protein_coding;protein_coding;protein_coding;protein_coding;protein_coding;processed_transcript;processed_transcript;processed_transcript",
    "name": "cg00903242",
    "note": "0.854812485786774",
    "position_to_tss": "-119;-294;-813;-306;-1167;-294;-1166;722;382;864",
    "refName": "chr3",
    "start": 196477,
    "transcript_id": "ENST00000256509.5;ENST00000397491.5;ENST00000421198.4;ENST00000427688.4;ENST00000453040.4;ENST00000481167.4;ENST00000489224.4;ENST00000444879.4;ENST00000451839.4;ENST00000452919.1",
    "uniqueId": "test-mbv-915",
  },
]
`;
